c10/core/TensorOptions.h

```diff
@@ -21,19 +21,19 @@ namespace c10 {
 DispatchKey computeDispatchKey(c10::optional<ScalarType> dtype, c10::optional<Layout> layout, c10::optional<Device> device);
 
 inline ScalarType dtype_or_default(c10::optional<ScalarType> dtype) {
-  return dtype.has_value() ? *dtype : get_default_dtype_as_scalartype();
+  return dtype.value_or_else(&get_default_dtype_as_scalartype);
 }
 
 inline caffe2::TypeMeta dtype_or_default(c10::optional<caffe2::TypeMeta> dtype) {
-  return dtype.has_value() ? *dtype : get_default_dtype();
+  return dtype.value_or_else(&get_default_dtype);
 }
 
 inline Layout layout_or_default(c10::optional<Layout> layout) {
-  return layout.has_value() ? *layout : kStrided;
+  return layout.value_or(kStrided);
 }
 
 inline Device device_or_default(c10::optional<Device> device) {
-  return device.has_value() ? *device : Device(kCPU);
+  return device.value_or_else([] {return Device(kCPU);});
```

Comment:
`value_or_else` here probably isn't buying you much

Reply:
you're probably right, the constructor is relatively cheap. But I wanted to be on the safe side and it definitely doesn't cost anything.